ember-headless-form/src/components/headless-form.ts

```diff
@@ -146,4 +201,21 @@ export default class HeadlessFormComponent<
   set<KEY extends FormKey<FormData<DATA>>>(key: KEY, value: DATA[KEY]): void {
     this.internalData[key] = value;
   }
+
+  @action
+  async handleFieldValidation(e: Event): Promise<void> {
+    const { target } = e;
+    const { name } = target as HTMLInputElement;
+
+    if (name) {
+      const field = this.fields.get(name as FormKey<FormData<DATA>>);
+
+      if (field) {
+        this.lastValidationResult = await this.validate();
+        field.validationEnabled = true;
+      }
+    } else {
+      // @todo how to handle custom controls that don't emit focusout/change events from native form controls?
```

Comment:
And yeah, that's the downside of that approach 😬

Reply:
imo,
custom controls can / should emit these events tho. 

"To be a custom control, it must behave like a control, including having <these events>"

Reply:
> imo,
> custom controls can / should emit these events tho.

Yeah, I agree.  We definitely will want validation to trigger for controls like `<PowerSelect`, `<PowerCalendar`, etc. used inside of forms.  Getting these elements to act more like their native versions would be 💯

Reply:
Ok, but the question is still whether to do 1. or 3. then.

For 1. to work, the event must have `event.target.name === @name passed to field`. A `<PowerSelect>` could have `<input type="hidden" name={{@name}}>` (does it, idk?), and if that triggers `focusout` then we are fine. 

But what about the example with the 3-select date picker. It will receive `@name="date"`, but will split this up internally to 3 selects of names `date_day`, `date_month`, `date_year`. When one of those selects triggers a `focusout`, our form will not know what field this belongs to, because `event.target.name === 'date'`. 

For 3., `event.target.name` would not matter, and everything would "just work"(TM). But the downside (wrapper div) remains...

Reply:
I think the caveat for 1 is fine.
For 3, can you write out the details of this approach by extending [this demo](https://limber.glimdown.com/edit?format=glimdown&t=%23%20Forms%0A%0A%23%23%23%20Using%20_The%20Platform_%0A%0AVanilla%20JavaScript%20has%20everything%20we%20need%20to%20handle%20form%20data%2C%20de-sync%20it%20from%20our%20source%20data%20and%20collect%20all%20user%20input%20upon%20submission.%0A%0AOther%20abstractions%2C%20such%20as%20the%20%22changeset%22%20concept%20contain%20a%20lot%20of%20this%20functionality%20and%20have%20additional%20utilities%20such%20as%20rollback%2C%20snapshots%2C%20forking%2C%20etc%2C%20but%20that%20is%20a%20topic%20for%20another%20demo.%0A%0AIn%20the%20form%20below%2C%20we%20create%20a%20Vanilla%E2%84%A2%20%5BHTML%20form%5D%5B2%5D%2C%20and%20only%20add%20%22Ember%22%20code%20for%20handling%20the%20form%20submission%20and%20field%20inputs.%20By%20default%2C%20form%20submissions%20will%20cause%20a%20page%20reload%2C%20so%20in%20a%20single-page-app%2C%20we%20need%20to%20prevent%20that%20default%20behavior.%0A%0AUsing%20the%20native%20API%2C%20%5BFormData%5D%5B1%5D%2C%20we%20can%20gather%20the%20user%20inputs%20when%20the%20user%20presses%20the%20submit%20button.%0A%0A%60%60%60gjs%20live%0Aimport%20%7B%20on%20%7D%20from%20%27%40ember%2Fmodifier%27%3B%0Aimport%20%7B%20cell%20%7D%20from%20%27ember-resources%27%3B%0A%0Aconst%20state%20%3D%20cell()%3B%0A%0Aconst%20handleInput%20%3D%20(event)%20%3D%3E%20%7B%0A%20%20let%20formData%20%3D%20new%20FormData(event.currentTarget)%3B%0A%20%20let%20data%20%3D%20Object.fromEntries(formData.entries())%3B%0A%0A%20%20state.current%20%3D%20JSON.stringify(data%2C%20null%2C%202)%3B%0A%7D%3B%0A%0Aconst%20handleSubmit%20%3D%20(%20event)%20%3D%3E%20%7B%0A%20%20event.preventDefault()%3B%0A%20%20handleInput(event)%3B%0A%7D%3B%0A%0A%3Ctemplate%3E%0A%20%20%3Cform%0A%20%20%20%20%7B%7Bon%20%27input%27%20handleInput%7D%7D%0A%20%20%20%20%7B%7Bon%20%27submit%27%20handleSubmit%7D%7D%0A%20%20%20%20class%3D%22grid%20gap-2%22%0A%20%20%20%20style%3D%22max-width%3A%20300px%22%0A%20%20%3E%0A%20%20%20%20%3Clabel%3E%20First%20Name%0A%20%20%20%20%20%20%3Cinput%20name%3D%27firstName%27%3E%0A%20%20%20%20%3C%2Flabel%3E%0A%0A%20%20%20%20%3Clabel%3E%20Favorite%20Date%0A%20%20%20%20%20%20%3Cinput%20type%3D%27date%27%20name%3D%27favoriteDate%27%3E%0A%20%20%20%20%3C%2Flabel%3E%0A%0A%20%20%20%20%3Cbutton%20type%3D%27submit%27%3ESubmit%3C%2Fbutton%3E%0A%20%20%3C%2Fform%3E%0A%0A%20%20%3Cbr%3E%3Cbr%3E%0A%0A%20%20FormData%3A%0A%20%20%3Cpre%3E%7B%7Bstate.current%7D%7D%3C%2Fpre%3E%0A%0A%20%20%3Cstyle%3E%0A%20%20%20%20input%20%7B%20border%3A%201px%20solid%3B%20%7D%0A%20%20%3C%2Fstyle%3E%0A%3C%2Ftemplate%3E%0A%60%60%60%0A%0A%3Chr%3E%0A%0A%23%23%23%20%22Platform%22%20References%0A%20-%20%5B%60%3Cform%3E%60%20on%20MDN%5D%5B2%5D%0A%20-%20%5B%60FormData%60%20on%20MDN%5D%5B1%5D%0A%0A%23%23%23%20Ember%20References%0A%20-%20%5Bthe%20%60on%60%20modifier%5D%5B3%5D%0A%0A%0A%5B1%5D%3A%20https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FAPI%2FFormData%0A%5B2%5D%3A%20https%3A%2F%2Fdeveloper.mozilla.org%2Fen-US%2Fdocs%2FWeb%2FHTML%2FElement%2Fform%0A%5B3%5D%3A%20https%3A%2F%2Fguides.emberjs.com%2Frelease%2Fcomponents%2Fcomponent-state-and-actions%2F%23toc_html-modifiers-and-actions%0A) -- i just want to be _super extra sure_ I understand how this wrapping div approach would work

Reply:
Not sure tbh how the demo would relate to this addon, but I'll try it this way:

The current [form.field template](https://github.com/CrowdStrike/ember-headless-form/blob/main/ember-headless-form/src/components/-private/field.hbs) would have a wrapping div like so:

```hbs
<div {{on "focusout (fn @triggerValidationFor @name)}} ...attributes>
  {{! here is what is currently in the template }}
</div>
```

`@triggerValidationFor` would be curried into the component when yielding it as a contextual component, so that's a private implementation detail.

From the caller's site:

```hbs
<HeadlessForm as |form|>
  <form.field @name="date" as |field|>  <-- this would now be a div that captures the event, and *knows* what @name it belongs to!
    <field.label>Date:</field.label>
    <CustomDatePickerComponentConsistingOfThreeSelectsWithUnknownNames
      @value={{field.value}}
      @onUpdate={{field.setValue}}
    />
    <field.errors/>
  </form
</HeadlessForm>
```

So here, as long as the custom component triggers a `focusout` event, validation would work ootb, no matter if there is no native form element, or what `name` it/they have or not have...

Btw, `triggerValidationFor` would _not_ call `e.stopPropagation()`, so the event would not stop to bubble at the `form.field` div level, if that was a concern. You could still attach event listeners to the `<form>` or even parent elements, and see the event...

Reply:
maybe this div is too much to assume. I imagine this may be something that a custom-control author would want to opt in to.

as in:
```hbs
<HeadlessForm as |form|>
  <form.field @name="date" as |field|> 
    <field.captureEvents @events={{"focusout"}}> or @events={{array "focusout" "change"}} ? idk
     ^-- this would now be a div that captures the event, and *knows* what @name it belongs to!
    <field.label>Date:</field.label>
      <CustomDatePickerComponentConsistingOfThreeSelectsWithUnknownNames
        @value={{field.value}}
        @onUpdate={{field.setValue}}
      />
    </field.captureEvents>
    <field.errors/>
  </form
</HeadlessForm>
```

But maybe that begins to defeat the purpose of the headless form?

>  would be curried into the component when yielding it as a contextual component, so that's a private implementation detail.

would there be a way to not add those event bindings if no validator is passed? (conditionally apply the on modifier?)

> Btw, triggerValidationFor would not call e.stopPropagation(), so the event would not stop to bubble at the form.field div level, if that was a concern

it was a concern! this is good news.

> You could still attach event listeners to the <form> or even parent elements, and see the event...

most excellent.

----------------------

I think I've now changed my opinion from being _all in_ on option 1, to _at least_ being a bit 50/50 on options 1 and 3 now.

I do have a concern though, so, what would happen if a field has multiple focusables? (this may sway me back to option 1? idk!)

```hbs
<HeadlessForm as |form|>
  <form.field @name="date" as |field|>
    <field.label>
       Date:
       <button {{on 'click' (fn this.helpAbout "date")>help</button>
        ^ - would we end up validating on events from here?
    </field.label>
    <CustomDatePickerComponentConsistingOfThreeSelectsWithUnknownNames
      @value={{field.value}}
      @onUpdate={{field.setValue}}
    />
    <field.errors/>
  </form
</HeadlessForm>
```